### PR TITLE
make typevar constraints work properly

### PR DESCRIPTION
**Pull request by KotlinIsland**
@jorenham want to review this?

### CODE REVIEW

File: test-data/unit/check-based-typevar.test [hunk 1]
```diff
@@ -0,0 +1,13 @@
+[case testConstrainedTypeVarWithVariance]
+from typing import Protocol, TypeVar
+
+T1 = TypeVar("T1", int, str)
+T2 = TypeVar("T2", int, str, covariant=True)  # E: TypeVar with constraints cannot have variance  [misc]
+T3 = TypeVar("T3", int, str, contravariant=True)  # E: TypeVar with constraints cannot have variance  [misc]
+T4 = TypeVar("T4", int, str, infer_variance=True)  # E: Unexpected argument to "TypeVar()": "infer_variance"  [misc] \
```
Comment (DetachHead) [inline test-data/unit/check-based-typevar.test:7]: `misc` already has way too many errors

Comment (KotlinIsland) [reply test-data/unit/check-based-typevar.test:7]: `valid-type` + #818

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,7 +1,8 @@
 # Basedmypy Changelog
 
 ## [Unreleased]
-
+### Added
+- make `TypeVar` constraints work properly (#817)
```
Comment (jorenham) [inline CHANGELOG.md:5]: "less improperly" 🤔 ?

**Review by jorenham — COMMENTED** [review]
I'm not very familiar with the mypy codebase; but this looks like an easy win to me. 
Perhaps it could help to quickly explain why this is illegal? Perhaps by noting that `TypeVar` with constraints isn't a type parameter at all, unlike a `TypeVar` without them. 

Maybe you could even go one step further; and disallow using these "bidirectional type mappings" (cus that's what they resemble most) for generic types altogether; and allow allow them in the context of functions. Because after all, they're just syntactic sugar for overloads with an equal input- and output type.

**Comment by jorenham** [comment]
It looks like the `scipy` primer diff are (almost?) all caused by a different union order? I thought unions were unordered 🤔 ?

File: mypy/checker.py [hunk 3]
```diff
@@ -2890,6 +2890,8 @@ def check_protocol_variance(self, defn: ClassDef) -> None:
             if not isinstance(tvar, TypeVarType):
                 # Variance of TypeVarTuple and ParamSpec is underspecified by PEPs.
                 continue
+            if tvar.values:
```
Comment (jorenham) [inline mypy/checker.py:2893]: so the `TypeVarTypes.values` are the constraints or something?

**Comment by KotlinIsland** [comment]
> It looks like the `scipy` primer diff are (almost?) all caused by a different union order? I thought unions were unordered 🤔 ?

epic fail moment, there's something somewhere that's using a `set` or something and makes the primer output diff all the time

File: mypy/checker.py (hunk 3 above)
Comment (KotlinIsland) [reply mypy/checker.py:2893]: yeah, apparently its legacy stuff

```py
T = TypeVar("T", values=(int, str))
```

Comment (jorenham) [reply mypy/checker.py:2893]: the name `constraints` makes no sense whatsoever, so I think I'd actually prefer `values` 🤷🏻

Comment (jorenham) [reply mypy/checker.py:2893]: why didn't they just name it `choices` or something?

**Comment by KotlinIsland** [comment]
> Maybe you could even go one step further; and disallow using these ... for generic types altogether; and allow allow them in the context of functions.

why? they aren't unsound or anything, they provide an mechanism to overload the entire class

**Comment by KotlinIsland** [comment]
> Perhaps it could help to quickly explain why this is illegal? Perhaps by noting that TypeVar with constraints isn't a type parameter at all, unlike a TypeVar without them.

so currently mypy does report on type annotations, but not type applications, and it currently doesn't provide any info, what would you suggest here?

```py
class A[T: (object, None)]: ...
A[int]()  # E: Value of type variable "T" of "A" cannot be "int"  [type-var]
             # N: this type parameter is not generic, it is constrained 
```

**Comment by KotlinIsland** [comment]
> ```python
> class A[T: (int, str)]: ...
> 
> 
> type BoolA[T: bool] = A[T]  # violates typing axiom #1: types don't change
> ```
> 
> edit: so in the same way that a "typevar with constraints" can't be covariant or contravariant, it also can't be invariant

in this case you would just get:
```
error: Value of type variable "T" of "A" cannot be "T"  [type-var]
```

**Comment by jorenham** [comment]
> > ```python
> > class A[T: (int, str)]: ...
> > 
> > 
> > type BoolA[T: bool] = A[T]  # violates typing axiom #1: types don't change
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > edit: so in the same way that a "typevar with constraints" can't be covariant or contravariant, it also can't be invariant
> 
> in this case you would just get:
> 
> ```
> error: Value of type variable "T" of "A" cannot be "T"  [type-var]
> ```

Ok that makes sense.

---

and how about

```pyi
class Gets[T: int | str]:
    def get(self) -> T: ...

class A[T: (int, str)](Gets[T]): ...
```

or 

```
class A[T: (int, str)]:
    def get(self) -> T: ...

class B[T: int](A[T]): ...
```

**Comment by KotlinIsland** [comment]
> and how about
> 
> ```python
> class Gets[T: int | str]:
>     def get(self) -> T: ...
> 
> class A[T: (int, str)](Gets[T]): ...
> ```

this is fine, i don't see any problem with it, it's the same as

```py
class Gets[T: int | str]:
    def get(self) -> T: ...

class A_int(Gets[int]): ...
class A_str(Gets[str]): ...
```

> ```python
> class A[T: (int, str)]:
>     def get(self) -> T: ...
> 
> class B[T: int](A[T]): ...
> ```

this is invalid:
```py
class B[T: int](A[T]): ...  # [E:](error: Value of type variable "T" of "A" cannot be "T"  [type-var]) 
```

File: mypy/typeshed/stdlib/builtins.pyi [hunk 4]
```diff
@@ -40,6 +39,7 @@ from typing import (  # noqa: Y022
     _NamedCallable,
     IO,
     Any,
+    AnyStr,
```
Comment (jorenham) [inline mypy/typeshed/stdlib/builtins.pyi:42]: Why not `TypeVar("_PathT_co", bound=str | bytes, covariant=True)`?

Comment (KotlinIsland) [reply mypy/typeshed/stdlib/builtins.pyi:42]: uhh, i don't really know anything about it, but i think `PathLike` is either `str` or `bytes`, never both?

Comment (jorenham) [reply mypy/typeshed/stdlib/builtins.pyi:42]: True, but now there's no way to type `PathLike[str | bytes]` if you don't care about the input, forcing you to use `Any`.

Comment (jorenham) [reply mypy/typeshed/stdlib/builtins.pyi:42]: A disjoint union is what would've been best here: https://en.wikipedia.org/wiki/Tagged_union

File: docs/source/based_features.rst [hunk 5]
```diff
@@ -134,6 +134,27 @@ So you are able to have functions with polymorphic generic parameters:
...
+
+    G = TypeVar("G", int, str)
+    class P(Protocol[G]):  # mypy reports an invalid error here
+        def f() -> E: ...
+
+    class A[T: (object, str)]: ...
+    a = A[int]()  # mypy doesn't report the error here
+
+    class B[T: int]: ...
+    type C = B[object]  # mypy doesn't report the error here
```
Comment (DetachHead) [inline docs/source/based_features.rst:155]: should briefly explain why its not valid to use variance on constraints

File: mypy/applytype.py [hunk 6]
```diff
@@ -31,11 +31,17 @@
 )
 
 
+class ReportIncompatibleTypevarValue(Protocol):
```
Comment (DetachHead) [inline mypy/applytype.py:34]: nit
```suggestion
class ReportIncompatibleTypeVarValue(Protocol):
```

File: mypy/checkexpr.py [hunk 7]
```diff
@@ -5124,6 +5124,15 @@ def apply_type_arguments_to_callable(
...
                 return AnyType(TypeOfAny.from_error)
+            for arg, tvar in zip(args, type_vars):
+                p = get_proper_type(arg)
+                if isinstance(p, AnyType):
+                    continue
+                if not isinstance(tvar, TypeVarType) or not tvar.values:
+                    continue
+                if not any(is_same_type(constraint, p) for constraint in tvar.values):
+                    # TODO: can we make the context the arg
+                    self.msg.incompatible_typevar_value(tp, arg, tvar.name, ctx, constrained=True)
```
Comment (DetachHead) [inline mypy/checkexpr.py:5135]: (should also have a better name for `p` but i left it because idk what it is)
```suggestion
            for arg, type_var in zip(args, type_vars):
                p = get_proper_type(arg)
                if isinstance(p, AnyType):
                    continue
                if not isinstance(tvar, TypeVarType) or not type_var.values:
                    continue
                if not any(is_same_type(constraint, p) for constraint in type_var.values):
                    # TODO: can we make the context the arg
                    self.msg.incompatible_typevar_value(tp, arg, type_var.name, ctx, constrained=True)
```

File: mypy/config_parser.py [hunk 8]
```diff
@@ -430,7 +430,7 @@ def destructure_overrides(toml_data: dict[str, Any]) -> dict[str, Any]:
                         raise ConfigTOMLValueError(
                             "toml config file contains "
                             "[[tool.mypy.overrides]] sections with conflicting "
-                            f"values. Module '{module}' has two different values for '{new_key}'"
+                            f"values. Module '{module}' has two different values for '{cast(str, new_key)}'"
```
Comment (DetachHead) [inline mypy/config_parser.py:433]: what else could it be

File: mypy/semanal_typeargs.py [hunk 9]
```diff
@@ -182,7 +186,7 @@ def validate_args(
                             is_error = True
                             self.fail(
                                 message_registry.INVALID_TYPEVAR_AS_TYPEARG.format(arg.name, name),
-                                ctx,
+                                ctx,  # TODO: can we use arg as the context
```
Comment (DetachHead) [inline mypy/semanal_typeargs.py:189]: what this means

File: mypy/test/data.py [hunk 10]
```diff
@@ -689,6 +690,11 @@ def split_test_cases(
                     name, parent.name, line_no
                 )
             )
+        version = m.group("version")
+        if version:
+            if sys.version_info < (3, int(version.split(".")[1])):
```
Comment (DetachHead) [inline mypy/test/data.py:695]: why not just compare againt the whole value
```suggestion
            if sys.version_info < (int(num) for num in version.split(".")):
```
or is there a function somewhere that already exists to compare versions as strings

File: mypy/typeshed/stdlib/builtins.pyi (hunk 4 above)
Comment (DetachHead) [reply mypy/typeshed/stdlib/builtins.pyi:42]: > True, but now there's no way to type `PathLike[str | bytes]` if you don't care about the input, forcing you to use `Any`.

`PathLike[str] | PathLike[bytes]`

File: test-data/unit/check-based-typevar.test [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+class A(Protocol[T1]):
+    """inferred as covariant, but it's constrained so it's valid"""
+    def f(self) -> T1:
+        pass
+
+
+[case testTypeVarWrongNameStillWorks]
+from typing import TypeVar
+
+T = TypeVar("NotT")  # E: String argument 1 "NotT" to TypeVar(...) does not match variable name "T"  [misc]
```
Comment (DetachHead) [inline test-data/unit/check-based-typevar.test:19]: is this test supposed to include a constraint and variance argument? otherwise it seems unrelated to this change

File: test-data/unit/check-based-typevar.test
Comment (DetachHead) [inline test-data/unit/check-based-typevar.test:1]: is there a test somewhere that makes sure you cant specify both constraints and bounds

File: test-data/unit/check-overloading.test [hunk 12]
```diff
@@ -5404,15 +5406,15 @@ def f_d(arg: str) -> None: ...
...
+Bad1 = TypeVar(Bad1)  # type: ignore
 
 def dec_e(f: Bad1) -> Bad1: ...  # type: ignore
 
 @overload
 def f_e(arg: int) -> None: ...
 @overload
 def f_e(arg: str) -> None: ...
-@dec_e  # E: Bad1? not callable
+@dec_e
```
Comment (DetachHead) [inline test-data/unit/check-overloading.test:5417]: whats this

Comment (KotlinIsland) [reply test-data/unit/check-overloading.test:5417]: there was a test for a decorator with an error in it, i made typevars not fail completely when the name is wrong, so i changed it, i think it's still a valid test though

File: test-data/unit/check-based-typevar.test
Comment (KotlinIsland) [reply test-data/unit/check-based-typevar.test:1]: yes

File: test-data/unit/check-based-typevar.test (hunk 11 above)
Comment (KotlinIsland) [reply test-data/unit/check-based-typevar.test:19]: it's just a small improvement to not break the typevar completely when the name is wrong

File: mypy/config_parser.py (hunk 8 above)
Comment (KotlinIsland) [reply mypy/config_parser.py:433]: `Any`

File: mypy/semanal_typeargs.py (hunk 9 above)
Comment (KotlinIsland) [reply mypy/semanal_typeargs.py:189]: the current error points to the type application, it would be more appropriate to point to the arg

File: mypy/test/data.py (hunk 10 above)
Comment (KotlinIsland) [reply mypy/test/data.py:695]: because it captures the `-3`

File: mypy/typeshed/stdlib/builtins.pyi (hunk 4 above)
Comment (KotlinIsland) [reply mypy/typeshed/stdlib/builtins.pyi:42]: > but now there's no way to type `PathLike[str | bytes]` 

not sure i follow, that's invalid and is in mypy too
https://mypy-play.net/?mypy=latest&python=3.12&gist=bcd7f8d13cb6c1678e3151f3900d1b46

also:

```py
PathLike[str] | PathLike[bytes]
```

https://github.com/KotlinIsland/basedmypy/pull/817#discussion_r1862851817

Comment (jorenham) [reply mypy/typeshed/stdlib/builtins.pyi:42]: > not sure i follow, that's invalid and is in mypy too https://mypy-play.net/?mypy=latest&python=3.12&gist=bcd7f8d13cb6c1678e3151f3900d1b46

Well, yea...
My point is that I don't think it should be illegal at all. Cus even though `str & bytes` isn't literally allowed, `str | bytes` is the best way we can currently express something that is "either `str` or `bytes`". It would be even better if there'd be a way to express `str ^ bytes` with which I mean `str | bytes - str & bytes`. But since we don't have that, I think we should use the next best *valid* alternative that we have at hand. And as I briefly explained earlier, I don't think that `T: (str, bytes)` is valid as a generic type parameter, as it is neither *co-*, *contra-*, nor *in*variant. Or on other words; it doesn't *vary* with the input; it *selects* that what is closest to it. So while that's valid when used as syntactic sugar for overloads, it shouldn't be confused with a type parameter. Because by doing so, you're dismissing the most fundamental and essential rule of type systems: Types don't change. But if you allow it anyway, and declare the type `PathLike[SubtypeOfStr]`, then it immediately changes into `PathLike[str]` -- Haram!

Comment (jorenham) [reply mypy/typeshed/stdlib/builtins.pyi:42]: > ```python
> PathLike[str] | PathLike[bytes]
> ```

That means that instead of

```pyi
def to_path[T: str | bytes](x: T, /) -> ToPath[T]: ...
```

you're must write it as

```pyi
@overload
def to_path(x: str, /) -> PathLike[str]: ...
@overload
def to_path(x: bytes, /) -> PathLike[bytes]: ...
@overload
def to_path(x: str | bytes, /) -> PathLike[str] | PathLike[bytes]: ...
```

So that defeats the purpose of the single use-case for which the "type-var with constraints" is intended in the first place; compressing overloads. Actually, it's even worse than that; it achieves the exact opposite.

Comment (KotlinIsland) [reply mypy/typeshed/stdlib/builtins.pyi:42]:
```py
def to_path[T: (str, bytes)](x: T, /) -> ToPath[T]: ...
```

Comment (jorenham) [reply mypy/typeshed/stdlib/builtins.pyi:42]: that doesn't include the `x: str | bytes` case

Comment (KotlinIsland) [reply mypy/typeshed/stdlib/builtins.pyi:42]: hmmm, I suppose a single value of `str | bytes` will always be `str` or `bytes`, so that's an improvement that could be made, but:

```py
def f[T: str|bytes](T1: T, T2: T) -> PathLike[T]:...
```
this would be really unsafe

Comment (jorenham) [reply mypy/typeshed/stdlib/builtins.pyi:42]: > this would be really unsafe

so `class PathLike[T: str | bytes](Protocol): ...` ?

Comment (KotlinIsland) [reply mypy/typeshed/stdlib/builtins.pyi:42]:
```py
def f[T: (str, bytes)](t: T) -> PathLike[T]:...

d: str | bytes
r = f(d)
reveal_type(r) # PathLike[str] | PathLike[bytes]
```
here we can see that the union distributes over the function and we get the desired behaviour

Comment (jorenham) [reply mypy/typeshed/stdlib/builtins.pyi:42]: > ```python
> def f[T: (str, bytes)](t: T) -> PathLike[T]:...
> 
> d: str | bytes
> r = f(d)
> reveal_type(r) # PathLike[str] | PathLike[bytes]
> ```
> 
> here we can see that the union distributes over the function and we get the desired behaviour

That doesn't work it pyright though 🤔

Comment (KotlinIsland) [reply mypy/typeshed/stdlib/builtins.pyi:42]: mmm, it's a limitation. I have an issue somewhere to do unions distribution properly, I think this might be a simple case that can be solved in a follow up

Comment (jorenham) [reply mypy/typeshed/stdlib/builtins.pyi:42]: > * [union distrubution of calls doesn't apply to constraints #831](https://github.com/KotlinIsland/basedmypy/issues/831)

As I explained in the issue, that's not valid, because it doesn't take the intersection type into account.
